pandas/core/indexes/base.py

```diff
@@ -2912,7 +2916,29 @@ def union(self, other, sort=None):
         self._assert_can_do_setop(other)
         other, result_name = self._convert_can_do_setop(other)
 
-        if self.dtype != other.dtype:
+        from pandas import CategoricalIndex
+
+        if isinstance(self, CategoricalIndex) and isinstance(other, CategoricalIndex):
```

Comment:
Can you check `isinstance(self.dtype, CategoricalDtype)` instead?

Reply:
Yes, of course I can.